src/main/java/io/wispforest/owo/serialization/SelfDescribedSerializer.java

```diff
@@ -0,0 +1,34 @@
+package io.wispforest.owo.serialization;
+
+import io.wispforest.owo.serialization.impl.SerializationAttribute;
+
+import java.util.Optional;
+import java.util.Set;
+
+public interface SelfDescribedSerializer<T> extends Serializer<T> {
+
+    @Override
+    default Set<SerializationAttribute> attributes(){
+        return Set.of(SerializationAttribute.SELF_DESCRIBING);
+    }
+
+    void empty();
+
+    @Override
+    default <V> void writeOptional(Codeck<V> codeck, Optional<V> optional) {
+        optional.ifPresentOrElse(v -> codeck.encode(this, v), this::empty);
```

Comment:
Don't `NbtEnd`s disappear after serialization, making deserialization use the default field value instead?

Reply:
Unknown, keeping such until fully confirm that it is not required as I don't want to just discard such just yet